Build process improvements

fivepiece opened the pull request:
* Remove openssl from build, try to use an OS provided package (or
uncomment for enabling the build)
* Use docker to build and run tests on multiple distributions

fivepiece commented:
Job 170.6 failed on the schedule test, should probably pass if you restart it

https://travis-ci.org/AdamISZ/joinmarket-clientserver/jobs/271127413#L2662-L2663

fivepiece commented:
Do not merge yet please.  I'm trying to re-apply this on my own fork but I think I broke my travis-ci instance somehow.

fivepiece commented:
travis issue fixed in my repo, closing again for now (sorry)

AdamISZ commented:
@fivepiece btw i can't restart travis jobs any more, it always just throws up an error when I try.

fivepiece commented:
@AdamISZ 
Globally or just this PR?
I'm reopening, I think I got it to do what I want now.

fivepiece commented:
Finally.  Some context for everthing:
This PR moves the build process from tarvis-ci's containers and into docker images.  This allows for testing multiple linux distros in one go.
When travis sets up a CI environment for a new push or PR, the following happens :
1. It first clones the current branch into a directory, then applies the changes from the pull \ PR.
2. We then exit the direcory containing this modified repository, and download bitcoin core into a `./downloads/` directory which we then cache for future runs (hopefully, will keep an eye on that).
3. Start the build matrix

You can see in the `.travis.yml` file, multiple environments are set, and a single command starting a build of a docker image which is a file from the pushed \ PR'ed repository we are currently testing/
The command in `script:` runs one time for each of the set environments, so eventually it tries to run each of the `Dockerfile.*` files.

The docker build, when run, will `COPY` both bitcoin core tarball and the `joinmarket-clientserver` directories into the docker images at runtime, then a complete build and test run is done within each of the images.

I've created docker files for all distros that I believe are widely used, and it should be easy to add more.  I locally built successfully on an Alpine Linux image.

The `install.sh` script itself still only handles cases of "sudo to install build deps" on ubuntu \ mint \ debian systems.  It's possible to add more, but should probably wait for actual requests because the differences between e.g centos\fedora are subtle yet significant. (centos needs the epel repo for example).

he `run_test.sh` file will run `corveralls` if we're in a docker build.  It's still broken as before, but I think that to make it work again I'll have to pass it some data like an API token, because it can't reach the one on travis-ci anymore.  Let me know if you want to do that.

It should be possible to use travis-ci's caching and docker's `COPY` to also not continually grab `libffi` and `libsodium`.  I'll look into that. It'll require some modifications to `install.sh`.
Currently I'm curious if the caching system works for the bitcoin core tarball.  I guess we'll be able to see that if you decide to merge

fivepiece commented:
- [x] Rbased
- [x] Modernized

Travis now runs both on its native setup and also runs the dockers.  Coveralls only runs on native linux build.
I removed jessie and replaced fedora {25,26} with 27 (release 28 will be added once https://bugzilla.redhat.com/show_bug.cgi?id=1574478 is fixed).

Downloading bitcoin core's x86_64 .tar.gz distribution is done as part of this PR, but it's also cached for subsequent runs, so runtimes should be faster (2-3 minutes).

fivepiece commented:
Rebased and :
- Trusty is now tested only on the host
- Core upgraded to version 0.16.1
